### PR TITLE
bpo-29110: Fix file object leak in `aifc.open`.

### PR DESCRIPTION
**Pull request by Uberi**
When given an invalid AIFF file, `aifc.open` doesn't clean up after itself. This patch makes it close the file object if it was opened by the library.

Also, it adds one unit test, `test.aifc_test.AifcMiscTest.test_close_opened_files_on_error`. The patch can be verified with `./python -m test -j0 test_aifc`.

### CODE REVIEW

File: Lib/test/test_aifc.py [hunk 1]
```diff
@@ -149,6 +149,16 @@ def test_skipunknown(self):
...
 
+    def test_close_opened_files_on_error(self):
+        non_aifc_file = findfile('pluck-pcm8.wav', subdir='audiodata')
+        with check_no_resource_warning(self):
+            try:
+                # Try opening a non-AIFC file, with the expectation
+                # that `aifc.open` will fail.
+                f = self.f = aifc.open(non_aifc_file, 'rb')
+            except:
+                pass
```
Comment (methane) [inline Lib/test/test_aifc.py]: Could you use `with self.assertRaises(aifc.Error):` instead of `try` ~ `except: pass`?

Comment (Uberi) [reply Lib/test/test_aifc.py]: Done!

File: Lib/test/test_aifc.py [hunk 2]
```diff
@@ -1,4 +1,4 @@
-from test.support import findfile, TESTFN, unlink
+from test.support import check_no_resource_warning, findfile, TESTFN, unlink
```
Comment (methane) [inline Lib/test/test_aifc.py:1]: Please wrap at 79 columns

File: Lib/test/test_aifc.py [hunk 3]
```diff
@@ -149,6 +149,14 @@ def test_skipunknown(self):
         #This file contains chunk types aifc doesn't recognize.
         self.f = aifc.open(findfile('Sine-1000Hz-300ms.aif'))
 
+    def test_close_opened_files_on_error(self):
+        non_aifc_file = findfile('pluck-pcm8.wav', subdir='audiodata')
+        with check_no_resource_warning(self):
+            with self.assertRaises(aifc.Error):
+                # Try opening a non-AIFC file, with the expectation
+                # that `aifc.open` will fail (without raising a ResourceWarning)
```
Comment (methane) [inline Lib/test/test_aifc.py]: same here.

**Review by methane — CHANGES_REQUESTED** [review]
Please add an entry at Lib section in Misc/NEWS.

File: Lib/test/test_aifc.py (hunk 2 above)
Comment (Uberi) [reply Lib/test/test_aifc.py:1]: It seems to be 77 characters over here. Is there something I'm missing?

**Comment by Uberi** [comment]
Rebased and squashed on upstream, added NEWS entry, and fixed the style thing noted above!

File: Lib/test/test_aifc.py (hunk 2 above)
Comment (methane) [reply Lib/test/test_aifc.py:1]: Uh, I'm sorry. no problem.

File: Misc/NEWS [hunk 4]
```diff
@@ -242,6 +242,9 @@ Library
 - bpo-29532: Altering a kwarg dictionary passed to functools.partial()
   no longer affects a partial object after creation.
 
+- bpo-29110: Fix file object leak in aifc.open() when file is given as a
+  filesystem path and is not in valid AIFF format.
+
```
Comment (methane) [inline Misc/NEWS]: Please add credit like `Patch by <your name>` at the end.

**Comment by Uberi** [comment]
Added and squashed!

**Review by methane — APPROVED** [review]

**Comment by methane** [comment]
We use "Squash and merge" in Github.  So squash is not needed for next time.
Normal commit make it easy to review changes from last commit.

BTW, since this is bugfix, would you create backport (cherry-pick) pull request for 3.5 and 3.6 branch?
I can do it for you, but "cherry pick + pull request + squash merge" workflow drops original author.

File: Lib/aifc.py [hunk 5]
```diff
@@ -543,16 +549,19 @@ class Aifc_write:
...
-            self._aifc = 0
+            file_object = builtins.open(f, 'wb')
+            try:
+                self.initfp(file_object)
+            except:
+                file_object.close()
+                raise
+
+            # treat .aiff file extensions as non-compressed audio
+            if f.endswith('.aiff'):
```
Comment (serhiy-storchaka) [inline Lib/aifc.py:560]: This is unrelated change and it introduces a regression. self._aifc is not defined if the file have non .aiff extension.

File: Lib/test/test_aifc.py [hunk 6]
```diff
@@ -149,6 +149,14 @@ def test_skipunknown(self):
         #This file contains chunk types aifc doesn't recognize.
         self.f = aifc.open(findfile('Sine-1000Hz-300ms.aif'))
 
+    def test_close_opened_files_on_error(self):
```
Comment (serhiy-storchaka) [inline Lib/test/test_aifc.py:152]: The test tests only Aifc_read. Needed a test for Aifc_write.

Is it possible to make tests common for aifc, sunau and wave?

File: Lib/aifc.py (hunk 5 above)
Comment (methane) [reply Lib/aifc.py:560]: `self.initfp(file_object)` initializes `self._aifc = 1`.

Comment (serhiy-storchaka) [reply Lib/aifc.py:560]: Ah, sorry, I missed this. It would be better to write just `self._aifc = not f.endswith('.aiff')` if you want to modernize this code.

But now it looks to me that `initfp()` never fails. All changes of `Aifc_write` were not needed.

Comment (methane) [reply Lib/aifc.py:560]: MemoryError and MaxRecursionError can happen, but that's very unlikely and hard to test.
That's why I didn't request testing Aifc_write.

File: Lib/test/test_aifc.py (hunk 6 above)
Comment (methane) [reply Lib/test/test_aifc.py:152]: #293 adds test for Aifc_write.

Since this type of test is depending on implementation (raise Exception on Edge case),
I didn't tried to make this generic.
